source/Irrlicht/CBAWMeshWriter.cpp

```diff
@@ -0,0 +1,269 @@
+
+			core::XXHash_256(data, m_headers[_headerIdx].blobSizeDecompr, m_headers[_headerIdx].blobHash);
+			_file->write(data, m_headers[_headerIdx].blobSizeDecompr);
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+
+			std::free(data);
+		}
+		template<>
+		void CBAWMeshWriter::exportAsBlob<core::ICPUBuffer>(core::ICPUBuffer* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			core::XXHash_256(_obj->getPointer(), _obj->getSize(), m_headers[_headerIdx].blobHash);
+			_file->write(_obj->getPointer(), _obj->getSize());
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+		}
+
+		bool CBAWMeshWriter::writeMesh(io::IWriteFile* _file, ICPUMesh* _mesh, int32_t _flags)
+		{
+			if (!_mesh || !_file)
+				return false;
+
+			const uint32_t FILE_HEADER_SIZE = 32;
+			_IRR_DEBUG_BREAK_IF(FILE_HEADER_SIZE != sizeof(core::BawFile::fileHeader))
+
+			_file->write(BAW_FILE_HEADER, FILE_HEADER_SIZE);
+
+			const uint32_t numOfInternalBlobs = genHeaders(_mesh);
+			const uint32_t OFFSETS_FILE_OFFSET = FILE_HEADER_SIZE + sizeof(uint32_t);
+			const uint32_t HEADERS_FILE_OFFSET = OFFSETS_FILE_OFFSET + numOfInternalBlobs*sizeof(m_offsets[0]);
+
+			m_offsets.set_used(numOfInternalBlobs);
+
+			_file->write(&numOfInternalBlobs, sizeof(numOfInternalBlobs));
+			// will be overwritten after actually calculating offsets
+			_file->write(m_offsets.const_pointer(), m_offsets.size() * sizeof(m_offsets[0]));
+
+			// will be overwritten after calculating not known yet data (hash and size for texture paths)
+			_file->write(m_headers.const_pointer(), m_headers.size() * sizeof(core::BlobHeader));
+
+			m_offsets.set_used(0); // set `used` to 0, to allow push starting from 0 index
+			for (int i = 0; i < m_headers.size(); ++i)
+			{
+				switch (m_headers[i].blobType)
+				{
+				case core::Blob::EBT_MESH_BUFFER:
+					exportAsBlob(reinterpret_cast<ICPUMeshBuffer*>(m_headers[i].handle), i, _file);
+					break;
+				case core::Blob::EBT_RAW_DATA_BUFFER:
+					exportAsBlob(reinterpret_cast<core::ICPUBuffer*>(m_headers[i].handle), i, _file);
+					break;
+				case core::Blob::EBT_DATA_FORMAT_DESC:
+					exportAsBlob(reinterpret_cast<IMeshDataFormatDesc<core::ICPUBuffer>*>(m_headers[i].handle), i, _file);
+					break;
+				case core::Blob::EBT_TEXTURE_PATH:
+					exportAsBlob(reinterpret_cast<video::IVirtualTexture*>(m_headers[i].handle), i, _file);
+					break;
+				}
+			}
+
+			const size_t prevPos = _file->getPos();
+
+			// overwrite offsets
+			_file->seek(OFFSETS_FILE_OFFSET);
+			_file->write(m_offsets.const_pointer(), m_offsets.size() * sizeof(m_offsets[0]));
+			// overwrite headers
+			_file->seek(HEADERS_FILE_OFFSET);
+			_file->write(m_headers.const_pointer(), m_headers.size() * sizeof(core::BlobHeader));
+
+			_file->seek(prevPos);
+
+			return true;
+		}
+
+		uint32_t CBAWMeshWriter::genHeaders(ICPUMesh * _mesh)
+		{
+			m_headers.clear();
```

Comment:
what about a blob for the _mesh itself!?

Reply:
Considered as not needed since it's just collection of ptrs to mesh buffers. Plus bounding box, but this can be recalculated after adding all loaded mesh buffers.

Reply:
Sure, but think what happens once we put more than one IMesh in a file.

P.S. The bounding box can be quite expensive to recalculate for a skinned mesh.